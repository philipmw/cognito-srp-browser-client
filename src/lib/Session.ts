import {
  HASH_TYPE,
  arrayToBase64,
  arrayToHexString,
  base64ToArray,
  hexStringToArray,
  padHex,
  stringToArray,
} from './util';
import { infoBits } from './constants';

export class Session {
  private hkdf: Uint8Array;
  private key: Uint8Array;

  constructor(poolname: string, username: string, key: Uint8Array, scramblingParameter: Uint8Array);
  constructor(poolname: string, username: string, hkdf: string);
  constructor(
    private poolname: string,
    private username: string,
    keyOrHkdf: Uint8Array | string,
    private scramblingParameter?: Uint8Array,
  ) {
    if (keyOrHkdf instanceof Uint8Array) {
      this.key = keyOrHkdf;
    } else {
      this.hkdf = hexStringToArray(keyOrHkdf);
    }
  }

  async calculateSignature(secretBlock: string, timestamp: string) {
    const poolArray = stringToArray(this.poolname);
    const userArray = stringToArray(this.username);
    const secretArray = base64ToArray(secretBlock);
    const timestampArray = stringToArray(timestamp);
    const dataArray = new Uint8Array(
      poolArray.length + userArray.length + secretArray.length + timestampArray.length,
    );
    dataArray.set(poolArray, 0);
    dataArray.set(userArray, poolArray.length);
    dataArray.set(secretArray, poolArray.length + userArray.length);
    dataArray.set(timestampArray, poolArray.length + userArray.length + secretArray.length);

    if (!this.hkdf) {
      this.hkdf = await this.calculateHkdf();
    }

    const hmacKey = await crypto.subtle.importKey(
      'raw',
      this.hkdf,
      { name: 'HMAC', hash: HASH_TYPE },
      true,
      ['sign'],
    );

    const hash = new Uint8Array(await crypto.subtle.sign('HMAC', hmacKey, dataArray));

    return arrayToBase64(hash);
  }

  async getHkdf() {
    if (!this.hkdf) {
      this.hkdf = await this.calculateHkdf();
    }

    return arrayToHexString(this.hkdf);
  }

  private async calculateHkdf() {
    if (!this.scramblingParameter) {
      throw new Error('No scrambing parameter set');
    }

    const scramblingKey = await crypto.subtle.importKey(
      'raw',
      hexStringToArray(padHex(this.scramblingParameter)),
      { name: 'HMAC', hash: HASH_TYPE },
      true,
      ['sign'],
    );

    const prk = new Uint8Array(
      await crypto.subtle.sign('HMAC', scramblingKey, hexStringToArray(padHex(this.key))),
    );

    const prkKey = await crypto.subtle.importKey(
      'raw',
      prk,
      { name: 'HMAC', hash: HASH_TYPE },
      true,
      ['sign'],
    );

    const hmac = new Uint8Array(await crypto.subtle.sign('HMAC', prkKey, stringToArray(infoBits)));

    return hmac.slice(0, 16);
  }
}
