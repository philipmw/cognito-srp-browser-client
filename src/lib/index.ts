export * from './ClientPasswordChallenge';
export * from './Session';
export * from './UserPool';
