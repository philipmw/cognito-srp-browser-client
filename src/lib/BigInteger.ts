import { BigInteger as _BigInteger } from 'jsbn';
import { arrayToHexString, hexStringToArray } from './util';

declare module 'jsbn' {
  export interface BigInteger {
    toArray(length?: number): Uint8Array;
  }
}

_BigInteger.prototype.toArray = function (length?: number) {
  let str = this.toString(16);

  if (length) {
    str = str.padStart(length * 2, '0');
  } else if (str.length % 2) {
    str = '0' + str;
  }

  return hexStringToArray(str);
};

export class BigInteger extends _BigInteger {
  static fromArray(array: Uint8Array) {
    return new BigInteger(arrayToHexString(array), 16);
  }
}
