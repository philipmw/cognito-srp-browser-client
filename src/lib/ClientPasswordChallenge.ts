import { ClientUser } from './UserPool';
import { BigInteger } from './BigInteger';
import {
  getBigInteger,
  calculateScramblingParameter,
  calculatePrivateKey,
  hexStringToArray,
} from './util';
import { g, N, Nbytes, getMultiplierParameter } from './constants';
import { Session } from './Session';

export class ClientPasswordChallenge {
  private a: BigInteger;
  private A: Uint8Array;

  constructor(
    public poolname: string,
    public user: ClientUser,
    a: string | Uint8Array,
  ) {
    this.a = getBigInteger(a);
  }

  calculateA() {
    if (!this.A) {
      this.A = g.modPow(this.a, N).toArray(Nbytes);
    }
    return this.A;
  }

  async getSession(B: string, salt: string) {
    const multiplierParameter = await getMultiplierParameter();
    const Bint = new BigInteger(B, 16);

    if (Bint.compareTo(BigInteger.ZERO) <= 0 || Bint.compareTo(N) >= 0) {
      throw new Error('A should be between 0 and N exclusive');
    }

    const privateKey = await calculatePrivateKey(this.poolname, this.user, salt);

    const scramblingParameter = await calculateScramblingParameter(
      this.calculateA(),
      hexStringToArray(B),
    );

    const sessionKey = Bint.subtract(multiplierParameter.multiply(g.modPow(privateKey, N)))
      .modPow(this.a.add(scramblingParameter.multiply(privateKey)), N)
      .mod(N)
      .toArray(Nbytes);

    return new Session(
      this.poolname,
      this.user.username,
      sessionKey,
      scramblingParameter.toArray(),
    );
  }
}
