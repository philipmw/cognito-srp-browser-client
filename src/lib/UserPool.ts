import { calculatePrivateKey, randomBytes, arrayToHexString } from './util';
import { N, Nbytes, g } from './constants';
import { ClientPasswordChallenge } from './ClientPasswordChallenge';

export interface ServerUser {
  username: string;
  salt: string;
  verifier: string;
}

export interface ClientUser {
  username: string;
  password: string;
}

export class UserPool {
  constructor(private poolname: string) {}

  async createUser(user: ClientUser, salt?: string): Promise<ServerUser> {
    if (!salt) {
      salt = arrayToHexString(randomBytes(16));
    }

    const privateKey = await calculatePrivateKey(this.poolname, user, salt);

    const verifier = arrayToHexString(g.modPow(privateKey, N).toArray(Nbytes));

    return { username: user.username, salt, verifier };
  }

  async getClientChallenge(user: ClientUser) {
    const a = randomBytes();
    return new ClientPasswordChallenge(this.poolname, user, a);
  }
}
