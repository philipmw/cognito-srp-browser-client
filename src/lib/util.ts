import { BigInteger } from './BigInteger';
import { ClientUser } from './UserPool';

export const HASH_TYPE = 'SHA-256';

export async function getHashString(data: Uint8Array | string) {
  const dataArray = data instanceof Uint8Array ? data : stringToArray(data);
  const hash = new Uint8Array(await crypto.subtle.digest(HASH_TYPE, dataArray));
  return arrayToHexString(hash).padStart(64, '0');
}

export function padHex(data: string | Uint8Array) {
  const hex = data instanceof Uint8Array ? arrayToHexString(data) : data;

  if (hex.length % 2) {
    return '0' + hex;
  } else if ('89ABCDEFabcdef'.includes(hex[0])) {
    return '00' + hex;
  } else {
    return hex;
  }
}

export function randomBytes(size = 32): Uint8Array {
  const data = new Uint8Array(size);
  crypto.getRandomValues(data);
  return data;
}

export async function calculateScramblingParameter(A: Uint8Array, B: Uint8Array) {
  const paddedA = hexStringToArray(padHex(A));
  const paddedB = hexStringToArray(padHex(B));

  const dataArray = new Uint8Array(paddedA.length + paddedB.length);
  dataArray.set(paddedA, 0);
  dataArray.set(paddedB, paddedA.length);

  const hash = new Uint8Array(await crypto.subtle.digest(HASH_TYPE, dataArray));

  return BigInteger.fromArray(hash);
}

export async function calculatePrivateKey(poolname: string, user: ClientUser, salt: string) {
  const hash = await getHashString(`${poolname}${user.username}:${user.password}`);
  const array = hexStringToArray(padHex(salt) + hash);
  return new BigInteger(await getHashString(array), 16);
}

export function getBigInteger(data: string | Uint8Array) {
  if (data instanceof Uint8Array) {
    return BigInteger.fromArray(data);
  } else {
    return new BigInteger(data, 16);
  }
}

export function arrayToHexString(data: Uint8Array): string {
  return Array.from(data)
    .map((b) => b.toString(16).padStart(2, '0'))
    .join('');
}

export function stringToArray(data: string): Uint8Array {
  const array = new Uint8Array(data.length);
  for (let n = 0; n < data.length; n++) {
    array[n] = data.charCodeAt(n);
  }
  return array;
}

export function hexStringToArray(data: string) {
  if (data.length % 2) {
    data = '0' + data;
  }
  const array = new Uint8Array(data.length / 2);
  for (let n = 0; n < data.length; n += 2) {
    const v = parseInt(data.substring(n, n + 2), 16);
    if (isNaN(v)) {
      throw new Error('Invalid value in hex string: ' + data.substring(n, 2));
    } else {
      array[n / 2] = v;
    }
  }
  return array;
}

export function base64ToArray(data: string): Uint8Array {
  return stringToArray(atob(data.replace(/\-/g, '+').replace(/_/g, '/')));
}

export function arrayToBase64(data: Uint8Array): string {
  return btoa(String.fromCharCode.apply(null, data));
}
