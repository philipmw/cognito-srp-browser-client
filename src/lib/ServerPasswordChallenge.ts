import { calculateScramblingParameter, getBigInteger, padHex } from './util';
import { BigInteger } from './BigInteger';
import { multiplierParameter, g, N, Nbytes } from './constants';
import { Session } from './Session';
import { ServerUser } from './UserPool';

export class ServerPasswordChallenge {
  private b: BigInteger;
  private B: Buffer;
  private verifier: BigInteger;

  constructor(
    public poolname: string,
    public user: ServerUser,
    b: string | Buffer,
  ) {
    this.verifier = new BigInteger(user.verifier, 16);
    this.b = getBigInteger(b);
  }

  calculateB() {
    if (!this.B) {
      this.B = multiplierParameter
        .multiply(this.verifier)
        .add(g.modPow(this.b, N))
        .mod(N)
        .toBuffer(Nbytes);
    }
    return this.B;
  }

  getSession(A: string) {
    A = padHex(A);
    const Aint = new BigInteger(A, 16);

    if (Aint.compareTo(BigInteger.ZERO) <= 0 || Aint.compareTo(N) >= 0) {
      throw new Error('A should be between 0 and N exclusive');
    }

    const scramblingParameter = calculateScramblingParameter(
      Buffer.from(A, 'hex'),
      this.calculateB(),
    );

    const sessionKey = Aint.multiply(this.verifier.modPow(scramblingParameter, N))
      .modPow(this.b, N)
      .mod(N)
      .toBuffer(Nbytes);

    return new Session(
      this.poolname,
      this.user.username,
      sessionKey,
      scramblingParameter.toBuffer(),
    );
  }
}
