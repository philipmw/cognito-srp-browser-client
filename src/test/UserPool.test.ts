import { UserPool, ClientPasswordChallenge } from '../lib/index';

import { salt, poolname, clientUser } from './constants';

describe('UserPool', () => {
  let pool: UserPool;

  beforeAll(() => {
    pool = new UserPool(poolname);
  });

  it('should create a user', async () => {
    const user = await pool.createUser(clientUser, salt);
    expect(user).toMatchSnapshot();
  });

  it('should provide client challenges', async () => {
    const challenge = await pool.getClientChallenge(clientUser);
    expect(challenge).toBeInstanceOf(ClientPasswordChallenge);
  });
});
