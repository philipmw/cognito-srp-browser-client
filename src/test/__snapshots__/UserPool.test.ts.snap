// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserPool should create a user 1`] = `
{
  "salt": "fc5f4d0bb754777037c53c901e87e0e6",
  "username": "testuser",
  "verifier": "8496d34afeac608ea6300f7937219eb33e1bcf5f0ad5c3e0e34abb35831358b18f92939bbe68667051847a9905a8d4bcd81a03dcc0b97af115b0fbe877c450073597b88ae39d04b0b87108f72fdab80c31418d010ba21440b53ac4bdaf08db8f1b39944e009b760f86bbc3d368a10c04758b693afbce20d1ba46002c496e9b48f6783d7acfd9e2ccfbd7a72eb8f615083a1f3a38be035ec2a96baa099854131c3b667a259f8391782dec1fcbab060962b30779903982006d9389b8ba384383441bac6e7ca5dc1980f8ab5981c46e4c58a8a3ca008aedb4feb015200d0fc64f6eb6378c77d59bfb27a4a1a7c454edf032a06f62ee99a905db0a2dccf129988a9613c12303be108c89d54e727a8bf4e40f1b90785bcaa392b982c9a75d02b458fc471ea1715fe6248a903e58cde4bf4ae18e9f4d3ce0413155d3b3ac7062d2f131121cad614d2481a3e71eb0d4a6ef0d8f1f6f3e796d18a36b2fb59a9d51ffa0aa114a12f59d7d6f9c6a7861a610a50d7b5e7b768847a2b8b9a5e89884448085ce",
}
`;
